src/app/SafeAttributePersistenceProvider.h

```diff
@@ -0,0 +1,253 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/ConcreteAttributePath.h>
+#include <app/data-model/Nullable.h>
+#include <app/util/attribute-metadata.h>
+#include <cstring>
+#include <inttypes.h>
+#include <lib/support/BufferReader.h>
+#include <lib/support/BufferWriter.h>
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * Interface for persisting attribute values. This will always write attributes in storage as little-endian
+ * and uses a different key to the AttributePersistenceProvider.
+ */
+
+class SafeAttributePersistenceProvider
+{
+public:
+    virtual ~SafeAttributePersistenceProvider() = default;
+    SafeAttributePersistenceProvider()          = default;
+
+    /**
+     * Write an attribute value from the attribute store (i.e. not a struct or
+     * list) to non-volatile memory.
+     *
+     * @param [in] aPath the attribute path for the data being written.
+     * @param [in] aValue the data to write.  Integers and floats are
+     *             represented in native endianness.  Strings are represented
+     *             as Pascal-style strings, as in ZCL, with a length prefix
+     *             whose size depends on the actual string type.  The length is
+     *             stored as little-endian.
+     *
+     *             Integer and float values have a size that matches the `size`
+     *             member of aMetadata.
+     *
+     *             String values have a size that corresponds to the actual size
+     *             of the data in the string (including the length prefix),
+     *             which is no larger than the `size` member of aMetadata.
+     */
+    virtual CHIP_ERROR SafeWriteValue(const ConcreteAttributePath & aPath, const ByteSpan & aValue) = 0;
```

Comment:
Is this really public API?  I would think this is a protected hook for actual implementations to implement, and the public API are the actual safe bits below.

And the documentation above about "native endian" is obviously wrong.  And for now strings are not supported at all, right?  And there is no aMetadata here.  (Looks like the docs on AttributePersistenceProvider were not updated when the aMetadata argument was removed from WriteValue....)

I think what we should consider doing here is just having these be protected hooks, documented as "just store this value as-is" without talking about how the value is encoded...  But if we really want we can document that the values are little-endian scalars.

Reply:
Yes, the documentation here was missed. It's been updated now. The SafeRead/WriteValue functions have been protected.

Reply:
.